src/bin/nydus-image/deduplicate.rs

```diff
@@ -0,0 +1,496 @@
+        let db = SqliteDatabase::new(db_path.strip_prefix('/').unwrap_or(db_path))?;
+        Ok(Self { sb, db })
+    }
+
+    /// Save metadata to the database: chunk and blob info.
+    pub fn save_metadata(&mut self, _mode: Option<&str>) -> anyhow::Result<Vec<Arc<BlobInfo>>> {
+        let tree = Tree::from_bootstrap(&self.sb, &mut ())
+            .context("Failed to load bootstrap for deduplication.")?;
+
+        // Create the blob table and chunk table.
+        self.db
+            .create_chunk()
+            .map_err(|e| anyhow!("Failed to create chunk: {:?}.", e))?;
+        self.db
+            .create_blob()
+            .map_err(|e| anyhow!("Failed to create blob: {:?}.", e))?;
+
+        // Save blob info to the blob table.
+        let blob_infos = self.sb.superblock.get_blob_infos();
+        for blob in &blob_infos {
+            self.db
+                .insert_blob(&BlobTable {
+                    blob_id: blob.blob_id().to_string(),
+                    blob_compressed_size: blob.compressed_size(),
+                    blob_uncompressed_size: blob.uncompressed_size(),
+                })
+                .map_err(|e| anyhow!("Failed to insert blob: {:?}.", e))?;
+        }
+
+        // Save chunk info to the chunk table.
+        let pre = &mut |t: &Tree| -> anyhow::Result<()> {
+            let node = t.lock_node();
+            for chunk in &node.chunks {
+                let index: u32 = chunk.inner.blob_index();
+                // Get the blob ID.
+                let chunk_blob_id = blob_infos[index as usize].blob_id();
+                // Insert the chunk into the chunk table.
+                self.db
+                    .insert_chunk(&ChunkTable {
+                        chunk_blob_id,
+                        chunk_digest: chunk.inner.id().to_string(),
+                        chunk_compressed_size: chunk.inner.compressed_size(),
+                        chunk_uncompressed_size: chunk.inner.uncompressed_size(),
+                        chunk_compressed_offset: chunk.inner.compressed_offset(),
+                        chunk_uncompressed_offset: chunk.inner.uncompressed_offset(),
+                    })
+                    .map_err(|e| anyhow!("Failed to insert chunk: {:?}.", e))?;
+            }
+            Ok(())
+        };
+        tree.walk_dfs_pre(pre)?;
+
+        Ok(self.sb.superblock.get_blob_infos())
+    }
+}
+
+pub trait Table<Conn, Err>: Sync + Send + Sized + 'static
+where
+    Err: Error + 'static,
+{
+    /// clear table.
+    fn clear(conn: &Conn) -> Result<(), Err>;
+
+    /// create table.
+    fn create(conn: &Conn) -> Result<(), Err>;
+
+    /// insert data.
+    fn insert(conn: &Conn, table: &Self) -> Result<(), Err>;
+
+    /// select all data.
+    fn list_all(conn: &Conn) -> Result<Vec<Self>, Err>;
+
+    /// select data with offset and limit.
+    fn list_paged(conn: &Conn, offset: i64, limit: i64) -> Result<Vec<Self>, Err>;
+}
+
+#[derive(Debug)]
+pub struct ChunkTable {
```

Comment:
How about:

```
struct ChunkTable {
    conn: Conn,
}

struct Chunk {
    chunk_blob_id: String,
    chunk_digest: String,
    chunk_compressed_size: u32,
    chunk_uncompressed_size: u32,
    chunk_compressed_offset: u64,
    chunk_uncompressed_offset: u64,
}
```

Reply:
We cannot make the Chunk and ChunkTable structs private. This is because the implementation of Deduplicate is designed to ensure the diversity of database types, specified as: pub struct Deduplicate<D: Database + Send + Sync>. Additionally, the methods within Deduplicate are interrelated with each other.